### ISSUE TITLE
[Subscriptions] - Design

### ISSUE THREAD

**Issue by deinok** [graphql-dotnet/graphql-client#25]
Enable the feature of sending subscriptions like query.

Open discusion about what is the best way to implement it, API, and technologies.

**Comment by deinok** [comment]
If using WebSockets there are two viable ways:
* https://github.com/sta/websocket-sharp & https://www.nuget.org/packages/WebSocketSharp
* System.Net.WebSockets.ClientWebSocket

**Comment by saba-sabrin** [comment]
Hi, I have tried the Subscription example to subscribe for a GraphQL subscription request. But, apparently, it is not receiving any notifications for the subscription. Could you please recheck the example and provide a concrete solution on how we can subscribe properly and listen to the incoming notifications? Would be really great and appreciated !! Looking forward to your reply. Thanks.

**Comment by deinok** [comment]
@saba-sabrin Todey i will create a sample

**Comment by saba-sabrin** [comment]
@deinok This is my code for using the subscription based on provided subscription usage:
ref:  graphql-client/samples/GraphQL.Client.Sample/Program.cs

I can't share the GraphQL API details as of it belongs to my client and I'm using their private server. The API also uses token based authentication. That's why I need an example of proper usage of subscription and notification reception.

This is my GraphQL query:

var Query = @"subscription {
                        deviceCharacteristic {
                            node {
                            characteristicType
                            characteristicValue
                            deviceId
                            deviceDisplayName
                        }
                    }
                }";

            // Simple delegate for checking "OnReceive" event data
            EventCheck eventCheckHandler = EventChecker;

// Function for subscription request using GraphQL Client
using (var graphQLHttpClient = new GraphQLHttpClient("ws://...."))  // Websocket based URL
            {
                var subscriptionResult = graphQLHttpClient.SendSubscribeAsync(Query).Result;
                subscriptionResult.OnReceive += (res) => {
                    Console.WriteLine("received event.");
                    eventCheckHandler(res);
                };
                
                Console.ReadKey();
            }
    
 // Delegate declaration
public delegate void EventCheck(GraphQLResponse res);

        public void EventChecker(GraphQLResponse res)
        {
            var response = res.Data;
            Console.WriteLine("Event detected");
        }

**Comment by deinok** [comment]
Try this and report back.
you set the event handler but the background thread that reads the websocket is not started
```csharp
            var subscriptionResult = graphQLHttpClient.SendSubscribeAsync(Query).Result;
            subscriptionResult.StartAsync();
            subscriptionResult.OnReceive += (res) => {
                Console.WriteLine("received event.");
                eventCheckHandler(res);
            };
```

**Comment by saba-sabrin** [comment]
![graphqlclient](https://user-images.githubusercontent.com/7774786/46318089-9a2ac400-c5d5-11e8-8985-6e27b0f8dd42.png)

@deinok As you can see in the screenshot, the "StartAsync" method is not found. I think, there are some ambiguity in the interface implementation in the class "GraphQLHttpSubscriptionResult". The "SendSubscribeAsync" method returns "IGraphQLSubscriptionResult" and this interface has no "StartAsync" method definition. Could you please have a look? Thanks.

**Comment by deinok** [comment]
Try cast to `GraphQLHttpSubscriptionResult` first what you have found is an error

**Comment by saba-sabrin** [comment]
I casted the response and tried with "StartAsync" method and this gives me an exception that, "The WebSocket has already been started". 

Could you please check the "SendSubscribeAsync()" method in the "GraphQLHttpClient" class? The "StartAsync" method has already been called there too! Any suggestions?

**Comment by deinok** [comment]
Hmmm not for now. Awaiting for this -> https://github.com/graphql-dotnet/server/issues/168
NVM, you server-backend, what language is using?

**Comment by saba-sabrin** [comment]
@deinok The GraphQL server backend is running in Golang.

[https://github.com/99designs/gqlgen](url)

I don't have much information about it other than the GraphQL related API conventions. Please inform when there are some subscription related updates. Thanks!

**Comment by saba-sabrin** [comment]
@deinok Any update on the subscription and notification event receiving part with the graphql-client library? Thanks.

**Comment by deinok** [comment]
I just need an open subscription server to run the tests

**Comment by deinok** [comment]
@rose-a Like a month ago I talked with the guys behind Rx.NET asking if for my case is better use events and delegates or use Reactive Extensions (Observables).
Seems like events do the same job. Mostly, the diference is returning an observable or an object with an event.

But I could be wrong, so if you have an idea of a better API, please share it with us.

**Comment by deinok** [comment]
@rose-a NETStandard1.3 is not a big issue, the library can be move to just NETStandard2.0 (This will also allow us simplificate some code)

Some general tips:
* The IMPLEMENTATION MUST use `ClientWebSocket`
* Code that "could be shared" with an implementation of a Server goes to `Common`, Client code to the `Client`
* The API Surface MUST dont care if the subscription is done via `ClientWebSocket` or any other way.
* The API Surface MUST don't care of the subscription implementation
* The IMPLEMENTATION can be the Apollo Subscription implementation (I'm not sure if there is a specification for that).
* client.Dispose() MUST close all `ClientWebSockets` directly or indirectly
* "IGraphQLSubscriptionResult.Dispose() or your equivalent" MUST close its `ClientWebSockets` directly or indirectly


And I think that I don't left anything. NVM, if you can create an API like that I will pay you a beer, I'm not joking

**Comment by deinok** [comment]
To anybody interested, we have created a branch called "subscriptions-api" in order to estabilize the API, but seems like using `System.Reactive` will be the way to go.
Also, many thanks to @rose-a for his work ;)

**Comment by rose-a** [comment]
@deinok I'm wondering if there is anything missing to the subscription API which prevents it to be merged into master (except from housekeeping)? Did you get around to review the current status of it?

**Comment by deinok** [comment]
@rose-a No, theres any problem with your subscription API. The main problem are some personal problems that makes me unable to fully review this. Finally this personal problems are solved and I'm back to the development.

The priority order is to move the library to use the nullable references, then add a local server to run the tests against, fix some issue mostly with authentication and then merge your API. So I expect that the next weekend we will be able to merge your API and finally release V2.0.0

**Comment by rose-a** [comment]
@deinok Regarding the local test server take a peek at the [IntegrationTestServer](https://github.com/graphql-dotnet/graphql-client/tree/subscriptions-api/tests/IntegrationTestServer) in the subscriptions-api branch...

**Comment by deinok** [comment]
Ty, thanks for the reminder

**Comment by jamespwright** [comment]
Is there any work that needs to be done on this still?  We wanted to use this for our integration test environment and discovered that the NuGet package version does not include the Subscription code.

**Comment by SKGGitHub** [comment]
What is the plan to incorporate Subscription Code in Stable version of this library?

**Comment by rose-a** [comment]
@deinok Any estimate when this might be incorporated into the big picture? If you need help somewhere don't mind to ask!

**Comment by SKGGitHub** [comment]
Hi Team, .NET Core 3.0 version is release now. There was information somewhere that GraphQL.Clinet stable version Subscription Code would be released in align with .NET Core 3.0.

Do we have any update on that please?

**Comment by ShearwaterJay** [comment]
@deinok Any info of when Subscriptions will be in?

**Comment by rose-a** [comment]
merged with #150
